### PR TITLE
perf: improve some instances in MonoidAlgebra

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ChrisHughes24** [comment]
!bench

File: Mathlib/Algebra/MonoidAlgebra/Basic.lean [hunk 1]
```diff
@@ -85,9 +85,14 @@ instance MonoidAlgebra.inhabited : Inhabited (MonoidAlgebra k G) :=
...
 
+instance MonoidAlgebra.addMonoid : AddMonoid (MonoidAlgebra k G) :=
+  inferInstanceAs (AddMonoid (G →₀ k))
+
 -- Porting note: The compiler couldn't derive this.
 instance MonoidAlgebra.addCommMonoid : AddCommMonoid (MonoidAlgebra k G) :=
-  inferInstanceAs (AddCommMonoid (G →₀ k))
+  --TODO: add reference to library note in PR #7432
+  { inferInstanceAs (AddCommMonoid (G →₀ k)) with
+    toAddMonoid := MonoidAlgebra.addMonoid _ _ }
```
Comment (mattrobball) [inline Mathlib/Algebra/MonoidAlgebra/Basic.lean:95]: Can you just do `add_comm := add_comm`?

File: Mathlib/Algebra/MonoidAlgebra/Basic.lean [hunk 2]
```diff
@@ -312,69 +319,105 @@ instance nontrivial [Semiring k] [Nontrivial k] [Nonempty G] : Nontrivial (Monoi
...
 
+instance neg [Ring k] : Neg (MonoidAlgebra k G) :=
+  Finsupp.neg
+
+instance sub [Ring k] : Sub (MonoidAlgebra k G) :=
+  Finsupp.sub
+
 instance addCommGroup [Ring k] : AddCommGroup (MonoidAlgebra k G) :=
-  Finsupp.addCommGroup
+  --TODO: add reference to library note in PR #7432
```
Comment (mattrobball) [inline Mathlib/Algebra/MonoidAlgebra/Basic.lean:336]: Should there be a `SubNegMonoid`?

File: Mathlib/Algebra/MonoidAlgebra/Basic.lean [hunk 3]
```diff
@@ -1189,9 +1234,14 @@ instance inhabited : Inhabited k[G] :=
...
 #align add_monoid_algebra.inhabited AddMonoidAlgebra.inhabited
 
+instance addMonoid : AddMonoid k[G] :=
+  inferInstanceAs (AddMonoid (G →₀ k))
+
 -- Porting note: The compiler couldn't derive this.
 instance addCommMonoid : AddCommMonoid k[G] :=
-  inferInstanceAs (AddCommMonoid (G →₀ k))
+  --TODO: add reference to library note in PR #7432
+  { inferInstanceAs (AddCommMonoid (G →₀ k)) with
```
Comment (mattrobball) [inline Mathlib/Algebra/MonoidAlgebra/Basic.lean:1243]: I feel like once you fix `FinSupp` instances this should be fine.

File: Mathlib/Algebra/MonoidAlgebra/Basic.lean [hunk 4]
```diff
@@ -1434,61 +1483,99 @@ instance nontrivial [Semiring k] [Nontrivial k] [Nonempty G] : Nontrivial k[G] :
...
+instance neg [Ring k] : Neg k[G] :=
+  Finsupp.neg
+
+instance sub [Ring k] : Sub k[G] :=
+  Finsupp.sub
+
 instance addCommGroup [Ring k] : AddCommGroup k[G] :=
-  Finsupp.addCommGroup
+  --TODO: add reference to library note in PR #7432
+  { Finsupp.addCommGroup with
```
Comment (mattrobball) [inline Mathlib/Algebra/MonoidAlgebra/Basic.lean:1501]: Similar here. This seems like it should be a problem with `Finsupp.addCommGroup` primarily.

**Review by mattrobball — COMMENTED** [review]
I think all the changes are fine but some seem to indicate we are working around performance issues from higher upstream, eg `Finsupp`. Can these be fixed there?

**Comment by mattrobball** [comment]
This should probably depend on #7430

**Comment by ChrisHughes24** [comment]
!bench

**Comment by ChrisHughes24** [comment]
!bench

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

**Pull request closed without merging**